### PR TITLE
Added Python article

### PR DESCRIPTION
**Pull request by Verdagon**

### CODE REVIEW

File: src/blog/python-data-races.vmd [hunk 1]
```diff
@@ -0,0 +1,146 @@
...
+Data races only happen in *parallel* code, not *concurrent* code, right? [# In Python, since only one thread can ever run at a time, it's concurrent. Concurrency is when we're doing multiple tasks at once, but only progressing one task at a time. Parallelism is when we e.g. use multiple cores to progress multiple threads at a time.]
+
+
+<slice />
+
+
+# A Python Data Race
+
+
+At first I said, "Yep, we still need mutexes to avoid race conditions, but the GIL makes Python immune to mere data races."
```
Comment (jhburns) [inline src/blog/python-data-races.vmd]: Quoting yourself if a bit strange,
> At first, I thought we still need mutexes to avoid race conditions, but the GIL makes Python immune to mere data races.

File: src/blog/python-data-races.vmd [hunk 2]
```diff
@@ -0,0 +1,146 @@
...
+
+
+...but that didn't seem right. I did a little bit of experimenting, and made [this program](https://gist.github.com/Verdagon/ef69b4c5e54a683ecd6a8afb715c3a80):
```
Comment (jhburns) [inline src/blog/python-data-races.vmd]: However, that...
Use proper grammar

File: src/blog/python-data-races.vmd [hunk 3]
```diff
@@ -0,0 +1,146 @@
...
+```
+
+
+
+Surprisingly, we didn't get 40000000, we got 31735072. [# This was on a 2.6 GHz 6-Core i7 Mac, for your computer to show a data race you may need to change the 400 or 100000.]
+
+
+This program's data race happens in the `counter = counter + 1` line. Let's pretend there are only two threads, and the program just started.
+
+ * Thread A loads 0 from `counter` into register A.
```
Comment (jhburns) [inline src/blog/python-data-races.vmd]: cpython doesn't actually use registers directly, it load variables onto stack frames that are stored on the heap according to this post https://tenthousandmeters.com/blog/python-behind-the-scenes-5-how-variables-are-implemented-in-cpython/. So change the terminology to "loads into the top of the thread's stack" or similar.

File: src/blog/python-data-races.vmd [hunk 4]
```diff
@@ -0,0 +1,146 @@
...
+[Go's map iteration order is random](https://news.ycombinator.com/item?id=7655948), in part because it prevents us from accidentally relying on iteration order. We could take some inspiration from that.
+
+
+I wonder if, in development mode, Python could use a shorter interval so that we notice any race conditions hiding in our code, and in release mode, could use this longer (5ms) interval. [# We could do this today, by using `sys.setswitchinterval()`!]
+
+
+This also relates to one's philosophy on determinism. We know that:
+
+ * Determinism helps a lot for reproducing bugs. Nobody likes heisenbugs. [# A "heisenbug" is when we encounter a bug, but when investigating, have difficulty making it happen again, because it's based on some non-deterministic factor.]
+ * Determinism can cause some bugs. If we accidentally depend on a hash map's ordering in 100 places, and then change our hash map's algorithm, suddenly we have 100 bugs.
```
Comment (jhburns) [inline src/blog/python-data-races.vmd]: Seems like this is phrased confusingly. It seems to imply the determinism is causing the bugs, but I'd say that user's reliance on determinism when they shouldn't be is the problem. So I'd rephrase to something like "Relying on determinism when it isn't fully supported by the language feature".

File: src/blog/python-data-races.vmd [hunk 5]
```diff
@@ -0,0 +1,146 @@
...
+
+
+These would seem to be incompatible, but a core goal of [Vale](https://vale.dev/) is to make races more obvious, and reproduce them easily.
+
+
+We could do this with *universal deterministic replayability*, where in development mode we record all non-deterministic inputs, such as command line arguments, stdin, sockets, files, etc. [# More specifically, it would just records any inputs from FFI into a file.] [# This feature is about halfway complete in Vale: the [FFI semantics](https://vale.dev/fearless) and FFI serialization are ready, but we don't record to a file yet.], plus the scheduling of all inter-thread messages and mutex lockings. This might seem difficult, but it's very possible when one uses a region-based borrow checker, and eliminates `unsafe` blocks and undefined behavior.
```
Comment (jhburns) [inline src/blog/python-data-races.vmd]: Really it sounds like replayability is possible due to the language being deterministic. Stuff like the "region-based borrow checker" is how Vale maintains determinism, but there are other ways to do it. I would clarify you mean Vale is deterministic because of these such features.

File: src/blog/python-data-races.vmd (hunk 3 above)
Comment (Verdagon) [reply src/blog/python-data-races.vmd]: I put it into a side-note, as those details might distract.

File: src/blog/python-data-races.vmd (hunk 4 above)
Comment (Verdagon) [reply src/blog/python-data-races.vmd]: Changed to "Relying on non-guaranteed determinism can cause some bugs.", hows that sound?

File: src/blog/python-data-races.vmd (hunk 5 above)
Comment (Verdagon) [reply src/blog/python-data-races.vmd]: Changed to "This might seem difficult, but it's possible for a language to guarantee determinism, for example if it uses a"
